pkg/controller/snapshot_controller.go

```diff
@@ -0,0 +1,766 @@
+		if err != nil {
+			glog.V(4).Infof("updating VolumeSnapshotContent[%s] error status: cannot update internal cache %v", newContent.Name, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (ctrl *csiSnapshotController) checkandUpdateSnapshotStatusOperation(snapshot *crdv1.VolumeSnapshot, content *crdv1.VolumeSnapshotContent) (*crdv1.VolumeSnapshot, error) {
+	status, _, err := ctrl.handler.GetSnapshotStatus(content)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check snapshot status %s with error %v", snapshot.Name, err)
+	}
+
+	newSnapshot, err := ctrl.updateSnapshotStatus(snapshot, status, time.Now(), IsSnapshotBound(snapshot, content))
+	if err != nil {
+		return nil, err
+	}
+	return newSnapshot, nil
+}
+
+// The function goes through the whole snapshot creation process.
+// 1. Trigger the snapshot through csi storage provider.
+// 2. Update VolumeSnapshot status with creationtimestamp information
+// 3. Create the VolumeSnapshotContent object with the snapshot id information.
+// 4. Bind the VolumeSnapshot and VolumeSnapshotContent object
+func (ctrl *csiSnapshotController) createSnapshotOperation(snapshot *crdv1.VolumeSnapshot) (*crdv1.VolumeSnapshot, error) {
+	glog.Infof("createSnapshot: Creating snapshot %s through the plugin ...", snapshotKey(snapshot))
+
+	class, err := ctrl.GetClassFromVolumeSnapshot(snapshot)
+	if err != nil {
+		glog.Errorf("createSnapshotOperation failed to getClassFromVolumeSnapshot %s", err)
+		return nil, err
+	}
+	volume, err := ctrl.getVolumeFromVolumeSnapshot(snapshot)
+	if err != nil {
+		glog.Errorf("createSnapshotOperation failed to get PersistentVolume object [%s]: Error: [%#v]", snapshot.Name, err)
+		return nil, err
+	}
+
+	// Create VolumeSnapshotContent name
+	contentName := GetSnapshotContentNameForSnapshot(snapshot)
+
+	// Resolve snapshotting secret credentials.
+	snapshotterSecretRef, err := GetSecretReference(class.Parameters, contentName, snapshot)
+	if err != nil {
+		return nil, err
+	}
+	snapshotterCredentials, err := GetCredentials(ctrl.client, snapshotterSecretRef)
+	if err != nil {
+		return nil, err
+	}
+
+	driverName, snapshotID, timestamp, csiSnapshotStatus, err := ctrl.handler.CreateSnapshot(snapshot, volume, class.Parameters, snapshotterCredentials)
+	if err != nil {
+		return nil, fmt.Errorf("failed to take snapshot of the volume, %s: %q", volume.Name, err)
+	}
+	glog.Infof("Create snapshot driver %s, snapshotId %s, timestamp %d, csiSnapshotStatus %v", driverName, snapshotID, timestamp, csiSnapshotStatus)
+
+	// Update snapshot status with timestamp
+	newSnapshot, err := ctrl.updateSnapshotStatus(snapshot, csiSnapshotStatus, time.Unix(0, timestamp), false)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create VolumeSnapshotContent in the database
+	volumeRef, err := ref.GetReference(scheme.Scheme, volume)
+
+	snapshotContent := &crdv1.VolumeSnapshotContent{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: contentName,
+		},
+		Spec: crdv1.VolumeSnapshotContentSpec{
```

Comment:
Please add `VolumeSnapshotClassName` in spec.
This will be used for getting secrets in `deleteSnapshotContentOperation`.

Reply:
I don't see storageClassName gets set in PV Spec in external-provisioner: https://github.com/kubernetes-csi/external-provisioner/blob/master/pkg/controller/controller.go#L410.  Do you know where that is set?

Reply:
I mean `crdv1.VolumeSnapshotContentSpec`

Reply:
I think you can get `storageClassName` in 'snapshot *crdv1.VolumeSnapshot`.

Reply:
I know.  My change is modeled after the similar change in external-provisioner.  I didn't find where storageClassName is set in PV spec there.  That's why I didn't set it in Content spec.

When you test delete volume, do you get secrets that you set in StorageClass?

Reply:
I got an error when I delete the snapshot, not the volume.
And I set `csiSnapshotSecretName` and `csiSnapshotSecretNamespace` in my `VolumeSnapshotClass`.
So I thought snapshot controller get secrets from `VolumeSnapshotClass`  for removing snapshots.
Why does the snapshot controller need storageClassName in PV for removing snapshot?

https://github.com/kubernetes-csi/external-snapshotter/blob/3e12fd6a36f694683ad628265da9b19998fd5878/pkg/controller/snapshot_controller.go#L573

Reply:
This is how external provisioner gets secrets at delete volume time.  See here: https://github.com/kubernetes-csi/external-provisioner/blob/master/pkg/controller/controller.go#L453

However, I don't see storageClassName gets set in PV spec when the volume is created in external provisioner.  So I wonder how secrets gets pass in there.  I'll check if it is set somewhere in the in-tree PV controller.

Reply:
I just add a line and it works.
```
		Spec: crdv1.VolumeSnapshotContentSpec{
			VolumeSnapshotRef: &v1.ObjectReference{
				Kind:       "VolumeSnapshot",
				Namespace:  snapshot.Namespace,
				Name:       snapshot.Name,
				UID:        snapshot.UID,
				APIVersion: "v1alpha1",
			},
			PersistentVolumeRef: volumeRef,
			VolumeSnapshotSource: crdv1.VolumeSnapshotSource{
				CSI: &crdv1.CSIVolumeSnapshotSource{
					Driver:         driverName,
					SnapshotHandle: snapshotID,
					CreatedAt:      timestamp,
				},
			},
                        // I added
			VolumeSnapshotClassName: snapshot.Spec.VolumeSnapshotClassName,
		},
	}
```
We can get secrets from `VolumeSnapshotClass` as well.
If I wrong, please let me know.

Reply:
Thanks for validating.  I still need to find out where it is set for PV spec.  Just want to make sure we are handling it consistently.

Reply:
I found out where this is set for PV spec.  It is set when PV is bound to PVC:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L1492

We should do the same thing here.  We can set VolumeSnapshotClass in the content spec when content is bound to snapshot in checkandBindSnapshotContent.  I'll fix that.

Reply:
@sngchlko This is fixed now.  Can you give a try?  So what you have changed is correct.  We still need that.

Reply:
Good!! It works. Thanks.

Reply:
Awesome:).  Thanks!